File: Problem2.py
# Time Complexity: O(N)
# Space Complexity: O(H) - H could be from logN for balanced tree to N for skewed tree

class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
      def helper(node, sum):
        if not node:
          return sum
        left = 0
        right = 0
        if node.left and node.right:  
          left = helper(node.left, sum * 10 + node.val)
          right = helper(node.right, sum * 10 + node.val)
          return left + right
        elif node.left:
          return helper(node.left, sum * 10 + node.val)
        else:
          return helper(node.right, sum * 10 + node.val)

      return helper(root, 0)
        

File: Problem2.js
//Problem2 (https://leetcode.com/problems/sum-root-to-leaf-numbers/)
// Acceted in leetcode
// time and space complexity is O(n)

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function(root) {

    let result =0;

const helper=(root, currentNum)=>{

    if(root === null) return;

    currentNum = currentNum*10 + root.val;

    if(root.left == null && root.right == null){
        result += currentNum;
    }

    helper(root.left, currentNum);
    helper(root.right, currentNum);

}
helper(root,0);
return result
}